plugin/src/main/java/org/opensearch/sql/plugin/catalog/CatalogServiceImpl.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.sql.catalog.CatalogService;
+import org.opensearch.sql.opensearch.security.SecurityAccess;
+import org.opensearch.sql.opensearch.setting.OpenSearchSettings;
+import org.opensearch.sql.storage.StorageEngine;
+
+/**
+ * This class manages catalogs and responsible for creating connectors to these catalogs.
+ */
+public class CatalogServiceImpl implements CatalogService {
+
+
+  private final Set<String> allowedConnectors = new HashSet<>() {{
+      add("prometheus");
+    }};
+
+  private Map<String, StorageEngine> storageEngineMap = new HashMap<>();
+
+  private static final CatalogServiceImpl INSTANCE = new CatalogServiceImpl();
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  public static CatalogServiceImpl getInstance() {
+    return INSTANCE;
+  }
+
+  private CatalogServiceImpl() {
+  }
+
+  /**
+   * This function reads settings and loads connectors to the data stores.
+   * This will be invoked during start up and also when settings are updated.
+   *
+   * @param settings settings.
+   */
+  public void loadConnectors(Settings settings) {
+    doPrivileged(() -> {
+      Boolean isPPLEnabled = (Boolean) OpenSearchSettings.PPL_ENABLED_SETTING.get(settings);
+      Boolean isFederationEnabled = CatalogSettings.FEDERATION_ENABLED.get(settings);
+      if (isPPLEnabled && isFederationEnabled) {
+        InputStream inputStream = CatalogSettings.CATALOG_CONFIG.get(settings);
+        if (inputStream != null) {
+          ObjectMapper objectMapper = new ObjectMapper();
+          objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+          try {
+            ArrayNode catalogs = objectMapper.readValue(inputStream, ArrayNode.class);
+            validateCatalogs(catalogs);
+            constructConnectors(catalogs);
+          } catch (IOException e) {
+            LOG.error("Catalog Configuration File uploaded is malformed. Verify and re-upload.");
+            throw new IllegalArgumentException(
+                "Malformed Catalog Configuration Json" + e.getMessage());
+          }
+        }
+      } else {
+        LOG.info("PPL federation is not enabled.");
+      }
+      return null;
+    });
+  }
+
+  @Override
+  public Optional<StorageEngine> getStorageEngine(String catalog) {
+    return Optional.ofNullable(storageEngineMap.get(catalog));
+  }
+
+  @Override
+  public Set<String> getCatalogs() {
+    return storageEngineMap.keySet();
+  }
+
+  private <T> T doPrivileged(PrivilegedExceptionAction<T> action) {
+    try {
+      return SecurityAccess.doPrivileged(action);
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to perform privileged action", e);
+    }
+  }
+
+  private StorageEngine createStorageEngine(JsonNode catalog) throws URISyntaxException {
```

Comment:
could we also maintain the OpenSearchStorageEngine in catalog service?

Reply:
We are using nodeClient for opensearchStorageEngine which is not available during catalogService initialization but only available during request response cycle.  Any other solutions would disrupt more code and don't want to do right now.

When we want to develop support for more than 1 opensearch instance would make the necessay changes and treat opensearch as yet another connector.